zcash_client_backend/src/data_api/wallet/input_selection.rs

```diff
@@ -598,6 +543,63 @@ impl<DbT: InputSource> InputSelector for GreedyInputSelector<DbT> {
 
+            #[cfg(not(feature = "transparent-inputs"))]
+            let ephemeral_balance = None;
+
+            #[cfg(feature = "transparent-inputs")]
+            let (ephemeral_balance, tr1_balance_opt) = {
+                if tr1_transparent_outputs.is_empty() {
+                    (None, None)
+                } else {
+                    // The ephemeral input going into transaction 1 must be able to pay that
+                    // transaction's fee, as well as the TEX address payments.
+
+                    // First compute the required total with an additional zero input,
+                    // catching the `InsufficientFunds` error to obtain the required amount
+                    // given the provided change strategy. Ignore the change memo in order
+                    // to avoid adding a change output.
+                    let tr1_required_input_value = match change_strategy
+                        .compute_balance::<_, DbT::NoteRef>(
+                            params,
+                            target_height,
+                            &[] as &[WalletTransparentOutput],
+                            &tr1_transparent_outputs,
+                            &sapling::EmptyBundleView,
+                            #[cfg(feature = "orchard")]
+                            &orchard_fees::EmptyBundleView,
+                            Some(&EphemeralBalance::Input(NonNegativeAmount::ZERO)),
+                            &wallet_meta,
+                        ) {
+                        Err(ChangeError::InsufficientFunds { required, .. }) => required,
+                        Err(ChangeError::DustInputs { .. }) => {
+                            unreachable!("no inputs were supplied")
+                        }
+                        Err(other) => return Err(InputSelectorError::Change(other)),
+                        Ok(_) => NonNegativeAmount::ZERO, // shouldn't happen
+                    };
+
+                    // Now recompute to obtain the `TransactionBalance` and verify that it
+                    // fully accounts for the required fees.
+                    let tr1_balance = change_strategy.compute_balance::<_, DbT::NoteRef>(
+                        params,
+                        target_height,
+                        &[] as &[WalletTransparentOutput],
+                        &tr1_transparent_outputs,
+                        &sapling::EmptyBundleView,
+                        #[cfg(feature = "orchard")]
+                        &orchard_fees::EmptyBundleView,
+                        Some(&EphemeralBalance::Input(tr1_required_input_value)),
+                        &wallet_meta,
+                    )?;
+                    assert_eq!(tr1_balance.total(), tr1_balance.fee_required());
+
+                    (
+                        Some(EphemeralBalance::Output(tr1_required_input_value)),
+                        Some(tr1_balance),
+                    )
+                }
+            };
```

Comment:
Why does this need to be inside the loop? `ephemeral_balance` and `tr1_balance_opt` should not need to be recomputed on each loop iteration, I think.

Reply:
Oh I see, it's because `wallet_meta` may change.  I don't especially like that this requires redundant computation for change strategies that don't need it, but I guess that's ok. It also increases the volume of code in an already-too-long loop (which could be fixed with a closure). I'll see if I can refactor this code to make it more understandable overall.